### DISCUSSION TITLE
[Help]: How to build an Embla Carousel Scrollbar

### DISCUSSION THREAD

**Discussion by imicadio** [davidjerleke/embla-carousel#1030] [Help] [answered]
### Summary

hi :) 

@davidjerleke your library is very good. I love this 😄 

I have "embla-carousel-react": "8.3.0" and I decided to make a scroll (scrollbar)

The scrollbar works quite nicely but I have a problem... or rather, my slider does not integrate with the arrows and the rest of the carousel if I move the slider: 

https://codesandbox.io/p/sandbox/embla-carousel-progress-react-forked-nqjjkt?workspaceId=d23f754a-ab87-4f51-b08d-bf0606a5023a 

I suppose that this piece of code is responsible for the potential errors and bugs that have occurred with me:
File: _src\js\CarouselScrollbar\CarouselScrollbar.tsx_
line of code: _77_

```
engine.animation.stop() 
engine.translate.to(rangeValueForEmblaEngine) 
engine.location.set(rangeValueForEmblaEngine) 
engine.index.set(closestIndex) 
```

Can I ask for help in solving the problem? 

https://github.com/user-attachments/assets/bb847692-8c4a-41b6-a9e6-6bb95afa2292

### If applicable, which variants of Embla Carousel are relevant to this question?

- [ ] embla-carousel **(Core)**
- [X] embla-carousel-react
- [ ] embla-carousel-vue
- [ ] embla-carousel-svelte
- [ ] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [ ] embla-carousel-class-names
- [ ] embla-carousel-fade
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

### CodeSandbox example

https://codesandbox.io/p/sandbox/embla-carousel-progress-react-forked-nqjjkt?workspaceId=d23f754a-ab87-4f51-b08d-bf0606a5023a

**Comment by TmLev** [comment]
Also interested in a custom scrollbar.

I took a look at your CodeSandbox example and it feels like there’s a state de-sync somewhere – the prev/next buttons aren’t picking up changes made via the scrollbar.  
Can we assume that the Embla engine itself has the correct internal state, since the slides are scrolling exactly as requested by the scrollbar?

**Reply by davidjerleke** [reply]
@TmLev yes, Embla has the correct state. You need to work with the internalEngine method and calculate the position using the limit property. Maybe I will have time to look into this after I’m done with v9.

**Reply by davidjerleke** [reply, marked as answer]
@imicadio @TmLev The CodeSandbox below is demonstrating the use of an `<input type="range">` as a scroll bar. This setup is just for illustrative purposes — feel free to swap in any input element that suits your needs. The core functionality lies in the `scrollToProgress` function, which handles the scroll logic.

- [CodeSandbox demo](https://5f6g63.csb.app/)
- [CodeSandbox code](https://codesandbox.io/p/sandbox/embla-carousel-drag-free-react-forked-5f6g63?file=%2Fsrc%2Fjs%2FEmblaCarousel.tsx%3A33%2C1-47%2C17)

**Reply by TmLev** [reply]
This is awesome – thank you so much for the demo and the code! Do you think it makes sense to add this to Embla as a plugin or as an example in the docs?

**Reply by davidjerleke** [reply]
Thanks for the suggestions, @TmLev! I’m glad you found the demo helpful. I’m not sure what would be best here — turning it into a plugin would require us to make assumptions about the scrollbar input, which isn’t ideal since different use cases might involve very different input elements or behavior. 

It might make more sense to include it as an example in the docs instead, since the `scrollToProgress` function is pretty minimal and easy to adapt.

**Reply by davidjerleke** [reply]
- Created issue #1218 and added it to the roadmap:
- #841.
